### PR TITLE
Updated Feign's version

### PR DESCRIPTION
**Pull request by GuiSim**
Updates the JAXRSContract and its test to be a copy of the JAXRSContract for 1.1 developed by OpenFeign.

Basically a copy of https://github.com/OpenFeign/feign/blob/master/jaxrs/src/main/java/feign/jaxrs/JAXRSContract.java but with a dependency on jax-rs 2.0.

Updated from Feign 9 to Feign 10.
This update is needed as Feign broke our JAXRSContract with 10.
See their fix here:
https://github.com/OpenFeign/feign/commit/8af587e6f5e42c84bb32eae69c45aaa2a9b595fb

 ```public Map<String, Collection<String>> queries() {
    Map<String, Collection<String>> queryMap = new LinkedHashMap<>();
    this.queries.forEach((key, queryTemplate) -> {
      List<String> values = new ArrayList<>(queryTemplate.getValues());

      /* add the expanded collection, but lock it */
      queryMap.put(key, Collections.unmodifiableList(values));
    });

    return Collections.unmodifiableMap(queryMap);
  }
```
From RequestTemplate 10.2.3

  ```public Map<String, Collection<String>> queries() {
    Map<String, Collection<String>> decoded = new LinkedHashMap<String, Collection<String>>();
    for (String field : queries.keySet()) {
      Collection<String> decodedValues = new ArrayList<String>();
      for (String value : valuesOrEmpty(queries, field)) {
        if (value != null) {
          decodedValues.add(urlDecode(value));
        } else {
          decodedValues.add(null);
        }
      }
      decoded.put(urlDecode(field), decodedValues);
    }
    return Collections.unmodifiableMap(decoded);
  }
```
From RequestTemplate 9.4.0

The new code made the array in the map's values immutable.

### CODE REVIEW

File: src/main/java/feign/jaxrs/JAXRSContract.java [hunk 1]
```diff
@@ -54,17 +53,21 @@ protected MethodMetadata parseAndValidateMetadata(Class<?> targetType, Method me
   @Override
   protected void processAnnotationOnClass(MethodMetadata data, Class<?> clz) {
     Path path = clz.getAnnotation(Path.class);
-    if (path != null) {
-      String pathValue = emptyToNull(path.value());
-      checkState(pathValue != null, "Path.value() was empty on type %s", clz.getName());
+    if (path != null && !path.value().isEmpty()) {
```
Comment (mapoulin) [inline src/main/java/feign/jaxrs/JAXRSContract.java]: Can path.value() be null?

File: src/main/java/feign/jaxrs/JAXRSContract.java [hunk 2]
```diff
@@ -54,17 +53,21 @@ protected MethodMetadata parseAndValidateMetadata(Class<?> targetType, Method me
...
-          pathValue = pathValue.substring(0, pathValue.length() - 1);
+        // Strip off any trailing slashes, since the template has already had slashes appropriately
+        // added
+        pathValue = pathValue.substring(0, pathValue.length() - 1);
       }
-      data.template().insert(0, pathValue);
+      // jax-rs allows whitespace around the param name, as well as an optional regex. The contract
+      // should
+      // strip these out appropriately.
+      pathValue = pathValue.replaceAll("\\{\\s*(.+?)\\s*(:.+?)?\\}", "\\{$1\\}");
```
Comment (mapoulin) [inline src/main/java/feign/jaxrs/JAXRSContract.java]: `.+?` isn't that the same as `.*` ?

File: pom.xml [hunk 3]
```diff
@@ -41,12 +41,12 @@
...
             <groupId>io.github.openfeign</groupId>
             <artifactId>feign-core</artifactId>
-            <version>9.4.0</version>
+            <version>10.2.3</version>
         </dependency>
         <dependency>
             <groupId>javax.ws.rs</groupId>
             <artifactId>javax.ws.rs-api</artifactId>
-            <version>2.0.1</version>
+            <version>2.1.1</version>
```
Comment (bbi22) [inline pom.xml]: Can we change for the new version of jax rs, [eclipse-ee4j/jaxrs-api](https://github.com/eclipse-ee4j/jaxrs-api)

File: src/main/java/feign/jaxrs/JAXRSContract.java (hunk 1 above)
Comment (GuiSim) [reply src/main/java/feign/jaxrs/JAXRSContract.java]: I don't know. I just copied from OpenFeign I don't want to change their code ;)

File: src/main/java/feign/jaxrs/JAXRSContract.java (hunk 2 above)
Comment (GuiSim) [reply src/main/java/feign/jaxrs/JAXRSContract.java]: Same thing, I didn't write this.

File: pom.xml (hunk 3 above)
Comment (GuiSim) [reply pom.xml]: This would break too many things, I'm sorry
image

**Pull request merged**